yaml.qmd

```diff
@@ -0,0 +1,874 @@
+
+1.  You get an error message, and the error message clearly points out your mistake.
+
+2.  You get an error message, but it's not clear from that message what or where your mistake is.
+
+3.  You don't get an error message at all.
+
+It's hard to enumerate all the ways you can make a mistake, but we'll give a couple of examples for each of these outcomes in the hope it helps you track down yours.
+
+Some examples in this chapter will result in an error from `quarto render`---we've highlighted these examples with a red border (TODO: what will we do in PDF), like this:
+
+``` {.yaml .error filename="render_error.qmd"}
+---
+title: Bad YAML : (
+---
+```
+
+The error messages shown in the text are current as of Quarto 1.6.
+
+We've split the examples by the kind of mistake you make:
+
+-   Using the wrong YAML syntax
+-   Using the wrong name for an option
+-   Using the wrong type of value for an option
+
+### Using the wrong YAML syntax
+
+When your YAML can't be parsed by Quarto you'll get a `YAMLException` error.
+These are actually surprisingly rare, not because everyone is great at YAML, but because most YAML syntax mistakes don't result in invalid YAML.
+The YAML remains valid, it just won't be interpreted in the way you intended.
+
+For example, if you forget the space after a colon when setting an option like `code-fold`:
+
+``` {.yaml .error filename="document.qmd"}
+---
+title: "Untitled"
+format:
+  html:
+    code-fold:true
+---
+```
+
+Then you don't get a `YAMLException` but instead a validation error:
+
+``` default
+Validation of YAML front matter failed.
+(line 5, columns 5--18) Field "html" has value code-fold:true, which must instead be an object
+
+✖ The value code-fold:true is a string.
+ℹ The error happened in location format:html.
+ℹ In YAML, key-value pairs in objects must be separated by a space.
+ℹ Did you mean code-fold: true instead?
+```
+
+The YAML has been parsed, but with the missing space `code-fold:true` was interpreted as a string not a mapping.
+Quarto does a good job here suggesting what your mistake might have been and a solution.
+
+Generally, if you get a validation error, make a careful check of your syntax first.
+
+### Specifying the wrong key
+
+There are some instances where when you use the wrong name for an option you'll get an error.
+For example, if you pluralize the `lightbox` option `loop`:
+
+``` {.yaml .error filename="document.qmd"}
+title: "Presentation"
+format: html
+lightbox: 
+  loops: true 
+```
+
+You'll get a YAML validation error that catches and reports the problem:
+
+``` default
+ERROR: Validation of YAML front matter failed.
+ERROR: In file long-string.qmd
+(line 5, columns 3--8) property name loops is invalid
+4: lightbox: 
+5:   loops: true 
+    ~~~~~~
+ℹ The error happened in location lightbox:loops.
+ℹ Did you mean loop?
+```
+
+Options where Quarto only allows a certain set of property names are known as *closed schemas*.
+I.e. the `lightbox` option has a closed schema.
```

Comment:
```suggestion
Options where Quarto only allows a certain set of property names are known as *closed schemas*, e.g., the `lightbox` option has a closed schema.
```

Reply:
Here a keyword/term is italicized. This goes back to my comment/question from earlier about how to mark these.